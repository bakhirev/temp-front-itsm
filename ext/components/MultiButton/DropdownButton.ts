import styled, { css } from 'styled-components';

import { DEFAULT_THEME } from '../common';
import { Button } from '../Button';

export const DropdownButton = styled(Button)`
  border-radius: ${({ theme }) => `0 ${theme.borderRadius} ${theme.borderRadius} 0`};
  ${({ kind }) =>
    kind === 'secondary' &&
    css`
      border-left: none;
      padding-left: 1px;
    `}
`;

DropdownButton.defaultProps = {
  theme: DEFAULT_THEME,
};
