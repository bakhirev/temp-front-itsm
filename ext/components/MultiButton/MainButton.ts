import styled, { css } from 'styled-components';

import { DEFAULT_THEME } from '../common';
import { Button } from '../Button';

export const MainButton = styled(Button)`
  ${({ kind, size = 'big', theme }) => css`
    border-radius: ${theme.borderRadius} 0 0 ${theme.borderRadius};
    position: relative;

    ${kind === 'secondary' &&
    css`
      border-right: none;
      padding-right: ${theme.button.paddingHorizontal[size]}px;
    `}
  `};
`;

MainButton.defaultProps = {
  theme: DEFAULT_THEME,
};
