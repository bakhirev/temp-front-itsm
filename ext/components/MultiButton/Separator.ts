import styled from 'styled-components';

import type { IMultiButtonProps } from './';

interface ISeparatorProps {
  size: IMultiButtonProps['size'];
}

const getSeparatorHeight = {
  big: '24px',
  medium: '24px',
  small: '20px',
  micro: '20px',
};

export const Separator = styled.div<ISeparatorProps>`
  position: relative;
  width: 1px;
  pointer-events: none;

  &::after {
    content: '';
    position: absolute;
    top: 50%;
    transform: translate(0, -50%);
    width: 1px;
    height: ${({ size = 'big' }) => getSeparatorHeight[size]};
  }
`;
