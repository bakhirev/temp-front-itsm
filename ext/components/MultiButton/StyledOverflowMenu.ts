import styled from 'styled-components';

import { OverflowMenu } from '../OverflowMenu';

export const StyledOverflowMenu = styled(OverflowMenu)`
  position: absolute;
  bottom: 0;
  right: 0;
  z-index: -1;
`;
