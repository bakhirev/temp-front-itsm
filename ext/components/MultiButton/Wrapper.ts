import styled from 'styled-components';

interface IWrapperProps {
  disabled?: boolean;
}

export const Wrapper = styled.div<IWrapperProps>`
  display: flex;
  position: relative;
  z-index: 1;
  ${({ disabled }) => disabled && 'pointer-events: none;'}
`;
