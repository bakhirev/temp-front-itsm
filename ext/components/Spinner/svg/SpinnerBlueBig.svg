<svg width="48" height="48" viewBox="0 0 48 48" fill="none" xmlns="http://www.w3.org/2000/svg">
<path fill-rule="evenodd" clip-rule="evenodd" d="M43.9553 37.3337C46.0346 34.2217 47.3495 30.6787 47.8128 26.9922C48.0193 25.3483 46.6569 24 45 24C43.3432 24 42.0262 25.3532 41.7532 26.9874C40.3302 35.5067 32.9234 42 24 42C14.0589 42 6.00001 33.9411 6.00001 24C6.00001 15.0766 12.4933 7.6698 21.0126 6.24674C22.6468 5.97377 24 4.65683 24 2.99997C24 1.34312 22.6517 -0.0193537 21.0077 0.187219C17.3213 0.650452 13.7783 1.96536 10.6663 4.0447C6.71955 6.68186 3.64341 10.4301 1.8269 14.8156C0.0103989 19.201 -0.464881 24.0266 0.461164 28.6821C1.38721 33.3377 3.67299 37.6141 7.02945 40.9705C10.3859 44.327 14.6623 46.6128 19.3178 47.5388C23.9734 48.4649 28.799 47.9896 33.1844 46.1731C37.5698 44.3566 41.3181 41.2804 43.9553 37.3337Z" fill="#0062FF"/>
</svg>
