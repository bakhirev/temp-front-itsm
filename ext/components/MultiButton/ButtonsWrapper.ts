import styled from 'styled-components';

import { DEFAULT_THEME } from '../common';
import { ITheme } from '../themes';

import { Separator } from './Separator';

export type Kind = 'primary' | 'secondary';

interface IButtonsWrapperProps {
  kind: Kind;
  disabled?: boolean;
}

export const getSeparatorBackgroundColor = (
  kind: Kind,
  theme: ITheme,
  disabled?: boolean,
  active?: boolean
) => {
  if (disabled) return kind === 'primary' ? theme.color.neutral[10] : 'transparent';
  if (active) return kind === 'primary' ? theme.color.primary[60] : 'transparent';
  return kind === 'primary' ? theme.color.primary[60] : 'transparent';
};

export const getSeparatorBorderColor = (
  kind: Kind,
  theme: ITheme,
  disabled?: boolean,
  active?: boolean
) => {
  if (disabled) return kind === 'primary' ? theme.color.neutral[10] : theme.color.neutral[30];
  if (active) return kind === 'primary' ? theme.color.primary[60] : 'none';
  return theme.color.primary[60];
};

export const getSeparatorAfterColor = (
  kind: Kind,
  theme: ITheme,
  disabled?: boolean,
  active?: boolean
) => {
  if (disabled) return theme.color.neutral[30];
  if (active) return kind === 'primary' ? theme.color.opacity.whiteMedium : theme.color.primary[60];
  return kind === 'primary' ? theme.color.opacity.whiteMedium : theme.color.primary[60];
};

export const ButtonsWrapper = styled.div<IButtonsWrapperProps>`
  display: flex;
  & > ${Separator} {
    background: ${({ kind, theme, disabled }) =>
      getSeparatorBackgroundColor(kind, theme, disabled)};
    border-top: 1px solid
      ${({ kind, theme, disabled }) => getSeparatorBorderColor(kind, theme, disabled)};
    border-bottom: 1px solid
      ${({ kind, theme, disabled }) => getSeparatorBorderColor(kind, theme, disabled)};
    &:after {
      background: ${({ kind, theme, disabled }) => getSeparatorAfterColor(kind, theme, disabled)};
    }
  }

  &:active {
    & ${Separator} {
      background: ${({ kind, theme, disabled }) =>
        getSeparatorBackgroundColor(kind, theme, disabled, true)};
      border-top: 1px solid
        ${({ theme, kind, disabled }) => getSeparatorBorderColor(kind, theme, disabled, true)};
      border-bottom: 1px solid
        ${({ theme, kind, disabled }) => getSeparatorBorderColor(kind, theme, disabled, true)};
      &:after {
        background: ${({ kind, theme, disabled }) =>
          getSeparatorAfterColor(kind, theme, disabled, true)};
      }
    }
  }
`;

ButtonsWrapper.defaultProps = {
  theme: DEFAULT_THEME,
};
