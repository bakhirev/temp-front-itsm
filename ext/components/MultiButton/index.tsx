import React, { FC, useState, useRef } from 'react';
import OutsideClickHandler from 'react-outside-click-handler';

import { ReactComponent as ChevronDownOutline } from '../Icons/system/ChevronDownOutline.svg';
import { ReactComponent as ChevronUpOutline } from '../Icons/system/ChevronUpOutline.svg';
import { KEY_CODES } from '../common';
import type { IButtonProps } from '../Button';
import type { IOverflowMenuProps, IOverflowMenuItem } from '../OverflowMenu';

import { MainButton } from './MainButton';
import { DropdownButton } from './DropdownButton';
import { StyledOverflowMenu } from './StyledOverflowMenu';
import { Wrapper } from './Wrapper';
import { ButtonsWrapper } from './ButtonsWrapper';
import type { Kind } from './ButtonsWrapper';
import { Separator } from './Separator';

const ICON_SIZE = 24;
const ICON_SIZE_SMALL = 20;

export interface IMultiButtonProps {
  /** Значение кнопки */
  button: IOverflowMenuItem;
  /** Массив опций для меню */
  list: IOverflowMenuProps['list'];
  /** Коллбэк на изменение выбранной опции (кнопки или пункта меню). Срабатывает при клике на опцию или при нажатии Enter/Space при фокусе на опции */
  onChange?: IOverflowMenuProps['onChange'];
  /** Внешний вид кнопки */
  kind?: Kind;
  /** Размер кнопки */
  size?: IButtonProps['size'];
  /** Отключение кнопки */
  disabled?: boolean;
  /** Имя класса для переопределения стилей */
  className?: string;
  /** data-test-id атрибут для тестирования компонента */
  dataTestId?: string;
}

export const MultiButton: FC<IMultiButtonProps> = ({
  className,
  dataTestId,
  disabled,
  kind = 'primary',
  size = 'big',
  button,
  list,
  onChange,
}) => {
  const overflowMenuRef: any = useRef(null);
  const wrapperRef = useRef<HTMLDivElement>(null);
  const [menuOpened, setMenuOpened] = useState<boolean>(false);
  const [value, setValue] = useState<IOverflowMenuItem>(button);

  const handleDropdownClick = () => {
    setMenuOpened(!menuOpened);
    overflowMenuRef.current?.select.focus();
  };
  const handleChange: IOverflowMenuProps['onChange'] = (value) => {
    setMenuOpened(false);
    setValue(value);
    onChange?.(value);
  };
  const handleOutsideClick = () => {
    setMenuOpened(false);
  };
  const handleKeyDown = (event: any) => {
    const { target } = event;
    let newFocusTarget;

    switch (event.keyCode) {
      case KEY_CODES.LEFT:
      case KEY_CODES.RIGHT:
        newFocusTarget =
          target !== target.parentNode.lastChild
            ? target.parentNode.lastChild
            : target.parentNode.firstChild;
        event.preventDefault();
        break;
      case KEY_CODES.ESCAPE:
        if (menuOpened) setMenuOpened(false);
        break;
      default:
        break;
    }

    if (newFocusTarget) {
      newFocusTarget.focus();
    }
  };

  const iconSize = size === 'big' || size === 'medium' ? ICON_SIZE : ICON_SIZE_SMALL;
  const Icon = menuOpened ? ChevronUpOutline : ChevronDownOutline;

  return (
    <OutsideClickHandler onOutsideClick={handleOutsideClick}>
      <Wrapper
        ref={wrapperRef}
        className={className}
        data-test-id={dataTestId}
        disabled={disabled}
        onKeyDown={handleKeyDown}
      >
        <ButtonsWrapper disabled={disabled} kind={kind}>
          <MainButton
            disabled={disabled}
            kind={kind}
            size={size}
            onClick={() => handleChange(button)}
          >
            {button.label}
          </MainButton>
          <Separator size={size} />
          <DropdownButton disabled={disabled} kind={kind} size={size} onClick={handleDropdownClick}>
            <Icon height={iconSize} width={iconSize} />
          </DropdownButton>
        </ButtonsWrapper>
        <StyledOverflowMenu
          disabled={disabled}
          hideDropdownIndicator
          innerRef={overflowMenuRef}
          list={list}
          menu={{
            opened: menuOpened,
            alignment: 'right',
          }}
          value={value}
          onChange={handleChange}
        />
      </Wrapper>
    </OutsideClickHandler>
  );
};
