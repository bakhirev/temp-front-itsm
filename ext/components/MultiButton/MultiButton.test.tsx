/* eslint-disable jsx-a11y/click-events-have-key-events */
/* eslint-disable jsx-a11y/no-static-element-interactions */
/* eslint-disable react/display-name */
import React from 'react';
import { shallow } from 'enzyme';

import { ReactComponent as ChevronDownOutline } from '../Icons/system/ChevronDownOutline.svg';
import { ReactComponent as ChevronUpOutline } from '../Icons/system/ChevronUpOutline.svg';
import { MultiButton } from '../MultiButton';
import type { IMultiButtonProps } from '../MultiButton';
import type { IOverflowMenuProps } from '../OverflowMenu';

jest.mock('../common', () => ({
  DEFAULT_THEME: {},
}));

jest.mock('../OverflowMenu', () => ({
  OverflowMenu: ({ defaultValue, list, value, onChange }: IOverflowMenuProps) => (
    <div onClick={() => onChange?.(list[2])} />
  ),
}));

jest.mock('../Icons/system/ChevronDownOutline.svg', () => ({
  ReactComponent: () => <div> down icon </div>,
}));

jest.mock('../Icons/system/ChevronUpOutline.svg', () => ({
  ReactComponent: () => <div> up icon </div>,
}));

describe('MultiButton', () => {
  const requiredProps: IMultiButtonProps = {
    list: [
      { label: 'foo', value: 2 },
      { label: 'bar', value: 3 },
      { label: 'baz', value: 4 },
    ],
    button: { label: 'zoo', value: 1 },
  };

  it('should render component with required props', () => {
    const wrapper = shallow(<MultiButton {...requiredProps} />);
    expect(wrapper).toMatchSnapshot();
  });

  it("should pass on button's props to Button component", () => {
    const buttonProps: IMultiButtonProps = {
      ...requiredProps,
      disabled: true,
      kind: 'secondary',
      size: 'small',
    };

    const wrapper = shallow(<MultiButton {...buttonProps} />);
    const button = wrapper.childAt(0).childAt(0).childAt(0);

    expect(button.prop('disabled')).toEqual(buttonProps.disabled);
    expect(button.prop('kind')).toEqual(buttonProps.kind);
    expect(button.prop('size')).toEqual(buttonProps.size);
  });

  it("should pass on menu's props to OverflowMenu component", () => {
    const menuProps = {
      ...requiredProps,
      disabled: true,
    };

    const wrapper = shallow(<MultiButton {...menuProps} />);
    const styledOverflowMenu = wrapper.childAt(0).childAt(1);

    expect(styledOverflowMenu.prop('list')).toEqual(menuProps.list);
    expect(styledOverflowMenu.prop('disabled')).toEqual(menuProps.disabled);
  });

  it('should render DropdownButton with arrow down icon by default', () => {
    const wrapper = shallow(<MultiButton {...requiredProps} />);
    const dropdownButton = wrapper.childAt(0).childAt(0).childAt(2);
    const children = dropdownButton.prop('children');

    expect(children).toEqual(<ChevronDownOutline height={24} width={24} />);
  });

  it('should render DropdownButton with arrow up icon when user clicks on it', () => {
    const wrapper = shallow(<MultiButton {...requiredProps} />);
    const dropdownButton = wrapper.childAt(0).childAt(0).childAt(2);

    dropdownButton.simulate('click');
    wrapper.update();

    const updatedButton = wrapper.childAt(0).childAt(0).childAt(2);
    const children = updatedButton.prop('children');

    expect(children).toEqual(<ChevronUpOutline height={24} width={24} />);
  });

  it('should open menu when user clicks on DropdownButton', () => {
    const wrapper = shallow(<MultiButton {...requiredProps} />);
    const dropdownButton = wrapper.childAt(0).childAt(0).childAt(2);
    const styledOverflowMenu = wrapper.childAt(0).childAt(1);
    expect(styledOverflowMenu.prop('menu').opened).toBeFalsy();

    dropdownButton.simulate('click');
    wrapper.update();

    const updatedStyledOverflowMenu = wrapper.childAt(0).childAt(1);
    expect(updatedStyledOverflowMenu.prop('menu').opened).toBeTruthy();
  });

  it('should call onChange with new value when user chooses one from the list', () => {
    const handleChange = jest.fn();
    const wrapper = shallow(<MultiButton {...requiredProps} onChange={handleChange} />);

    const overflowMenu = wrapper.childAt(0).childAt(1).dive().dive();
    overflowMenu.simulate('click');

    expect(handleChange).toBeCalledWith(requiredProps.list[2]);
  });
});
