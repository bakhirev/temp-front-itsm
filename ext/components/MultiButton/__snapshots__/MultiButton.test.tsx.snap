// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MultiButton should render component with required props 1`] = `
<OutsideClickHandler
  disabled={false}
  display="block"
  onOutsideClick={[Function]}
  useCapture={true}
>
  <styled.div
    onKeyDown={[Function]}
  >
    <styled.div
      kind="primary"
      theme={Object {}}
    >
      <Styled(Component)
        kind="primary"
        onClick={[Function]}
        size="big"
        theme={Object {}}
      >
        zoo
      </Styled(Component)>
      <styled.div
        size="big"
      />
      <Styled(Component)
        kind="primary"
        onClick={[Function]}
        size="big"
        theme={Object {}}
      >
        <ReactComponent
          height={24}
          width={24}
        />
      </Styled(Component)>
    </styled.div>
    <Styled(OverflowMenu)
      hideDropdownIndicator={true}
      innerRef={
        Object {
          "current": null,
        }
      }
      list={
        Array [
          Object {
            "label": "foo",
            "value": 2,
          },
          Object {
            "label": "bar",
            "value": 3,
          },
          Object {
            "label": "baz",
            "value": 4,
          },
        ]
      }
      menu={
        Object {
          "alignment": "right",
          "opened": false,
        }
      }
      onChange={[Function]}
      value={
        Object {
          "label": "zoo",
          "value": 1,
        }
      }
    />
  </styled.div>
</OutsideClickHandler>
`;
